compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -1722,38 +1720,43 @@ abstract class GenJSCode extends plugins.PluginComponent
           val sym = lhs.symbol
           if (sym.isStaticMember)
             abort(s"Assignment to static member ${sym.fullName} not supported")
-          val genRhs = genExpr(rhs)
-          lhs match {
-            case Select(qualifier, _) =>
-              val ctorAssignment = (
-                  currentMethodSym.isClassConstructor &&
-                  currentMethodSym.owner == qualifier.symbol &&
-                  qualifier.isInstanceOf[This]
-              )
-              if (!ctorAssignment && !suspectFieldMutable(sym))
-                unexpectedMutatedFields += sym
-
-              val genQual = genExpr(qualifier)
-
-              if (isScalaJSDefinedJSClass(sym.owner)) {
-                val genLhs = if (isExposed(sym))
-                  js.JSBracketSelect(genQual, js.StringLiteral(jsNameOf(sym)))
-                else
-                  js.JSDotSelect(genQual, encodeFieldSym(sym))
-                val boxedRhs =
-                  ensureBoxed(genRhs,
-                      enteringPhase(currentRun.posterasurePhase)(rhs.tpe))
-                js.Assign(genLhs, boxedRhs)
-              } else {
+          if (rhs.symbol == Runtime_optionalUndefined) {
```

Comment:
No, because we cannot distinguish from the *initial* assignment caused by the declaration of the field, versus other assignments elsewhere in the constructor (and methods). Consider:
```scala
@ScalaJSDefined
class Foo extends js.Object {
  @JSOptional var x: js.UndefOr[Int] = js.undefined
  x = 5
  println(x)
  x = js.undefined
  println(x)
}
```
would result in the following constructor definition:
```scala
def this() {
  this.x = js.undefined // needs to be dropped
  this.x = 5
  println(this.x)
  this.x = js.undefined // needs to be kept!
  println(this.x)
}
```